src/algorithms/semihdp_sampler.hpp

```diff
@@ -0,0 +1,205 @@
+#ifndef SRC_ALGORITHMS_SEMI_HDP_SAMPLER_HPP
+#define SRC_ALGORITHMS_SEMI_HDP_SAMPLER_HPP
+
+#include <omp.h>
+#include <proto/cpp/semihdp.pb.h>
+
+#include <Eigen/Dense>
+#include <numeric>
+#include <src/collectors/memory_collector.hpp>
+#include <src/hierarchies/base_hierarchy.hpp>
+#include <src/utils/distributions.hpp>
+#include <src/utils/rng.hpp>
+#include <stan/math/prim.hpp>
+#include <stdexcept>
+#include <vector>
+
+using bayesmix::MarginalState;
+using bayesmix::SemiHdpState;
+
+/*
+ * This class implements the algorithm for posterior simulation under the
+ * semi-hierarchical Dirichlet process in [1].
+ * 
+ * Extra goodies: we can now tune the pseudo-prior generation, by randomly
+ * drawing the cardinalities from a multinomial distribution, and perturbing
+ * the atoms of the mixing measure.
+ *
+ * [1]: "The semi-hierarchical Dirichlet Process and its application to
+ * clustering homogeneous distributions", Beraha, Guglielmi and Quintana
+ * arXiv: 2005.10287
+ */
+
+class SemiHdpSampler {
+ protected:
+  bayesmix::SemiHdpParams params;
+
+  std::vector<Eigen::MatrixXd> data;  // one vector per group
+  int ngroups;
+  std::vector<int> n_by_group;
+
+  std::shared_ptr<BaseHierarchy> G0_master_hierarchy;
+  std::shared_ptr<BaseHierarchy> G00_master_hierarchy;
+
+  // these should be just copies
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> rest_tables;
+  // idiosincratic tables
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> private_tables;
+  // shared tables
+  std::vector<std::shared_ptr<BaseHierarchy>> shared_tables;
+  // counts
+  std::vector<std::vector<int>> n_by_table;
+
+  // stuff for pseudoprior
+  std::vector<std::vector<std::shared_ptr<BaseHierarchy>>> rest_tables_pseudo;
+  std::vector<std::vector<int>> n_by_table_pseudo;
+
+  // theta_{ij} = theta_{c_i, s_{ij}}
+  std::vector<std::vector<int>> table_allocs;
+  // which restaurant each grups enters from
+  std::vector<int> rest_allocs;
+  std::vector<bool> is_used_rest;
+  Eigen::VectorXd omega;
+  std::vector<std::vector<int>> table_to_shared, table_to_private;
+
+  // number of theta_stars equal to tau_h across all groups
+  std::vector<int> m;
```

Comment:
The renamings are really nice, I just don't understand why you left these two out...

Reply:
I just ran out of names... :) I will think this harder